pytket/binders/circuit/pool.cpp

```diff
@@ -0,0 +1,196 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <pybind11/pybind11.h>
+
+#include "Circuit/CircPool.hpp"
+#include "binder_utils.hpp"
+#include "typecast.hpp"
+
+namespace py = pybind11;
+
+namespace tket {
+
+void init_pool(py::module &m) {
+  /* Circuit library */
+  py::module_ pool_m = m.def_submodule(
```

Comment:
Suggestion: keep these functions out of the API (by prepending an `_` and keeping them inofficial for example), and possibly expose python functions `get_singleq_decomp(singleqs: frozenset)` and a similar two-qubit version (which currently are essentially your dicts `_tk1_circs` and `_CX_CIRCS`). This way we are free to amend and augment the available decompositions as we like.

Or if we would like something more granular: we provide all these decompositions as part of some "circuit library", in which each decomposition has a string ID (its function name) and we expose `get_decomposition(decomp_id: str, *args)`, to retrieve them. We can then say explicitly that the exact range and behaviour of decompositions on offer is an implementation detail and that we recommend using `auto_rebase`. Sounds like a hack, but actually, from a user's perspective, I find that this is reasonable and makes more sense than just a huge dump of functions with limited documentation. Whoever is hell-bent on using a very specific decomposition can accept that it might not be available identically in future versions.

Unless @ss2165 you had something specific in mind when you were exposing these?

Reply:
I like the idea of a string based query for circuits, I propose a singleton library class which can report which circuits it has available and allow them to be requested. I will implement this in C++ to not need exposing all these random functions. If this proves cumbersome, @lmondada's first suggestion is still better than exposing all of them as part of the API.